lib/validators/built-in/invalid-links.js

```diff
@@ -0,0 +1,60 @@
+import _ from 'lodash';
+import {
+  forEachComponent, forEachField, labelComponent, getPreviewText
+} from '../helpers';
+
+export const label = 'Invalid Links',
+  description = 'There are invalid links on the page',
+  type = 'error';
+
+
+/**
+ * @param {string} url: The href value of a link.
+ * @returns {boolean} true if link is hash (same-page) link, else false.
+ */
+const isHash = (url) => (/^#/u).test(url);
+
+/**
+ * @param {string} url: The href value of a link.
+ * @returns {boolean} true if link is valid URL, else false.
+ */
+const isValid = (url) => {
+  try {
+    new URL(url);
+  } catch (err) {
+    return false;
+  }
+
+  return true;
+};
+
+/**
+ * Collects all invalid links from the page state and returns
+ * an array of validation errors.
+ *
+ * @param {Object} state
+ * @returns {Object[]} An array of validation errors.
+ */
+export function validate(state) {
+  const errors = [];
+
+  forEachComponent(state, (componentData, uri) => forEachField(state, componentData, uri, (field) => {
```

Comment:
this ends up combing over a lot of components and their properties. i wonder if it'd be worth whitelisting this only for generic clay components, e.g. `clay-paragraph` and `clay-subheader`. could also see an argument for this as is though, as a general all-purpose link validator

Reply:
I was just checking out `forEachComponent` and noticed that it accepts a filter.
I think limiting it to `clay-paragraph` could be a good idea (although `edit-links.js` does not).

Maybe an initial release w/ `clay-paragraph` as a filter (to reduce the blast area) and then that can be loosened if the case presents itself.

Reply:
I applied the filter for `clay-paragraph` and `clay-subheader`.

Reply:
fair enough. not a necessity for this PR, but there also things called [behaviors in kiln](https://github.com/clay/clay-kiln/blob/efe67eed4ff81f505f9ff109c1f27738583422d7/lib/core-data/behaviors2input.js) and maybe the best multi-purpose solution would to have link validation settable on the component schema level

Reply:
That's really good to know, I'm going to throw that into the PR description just to note.